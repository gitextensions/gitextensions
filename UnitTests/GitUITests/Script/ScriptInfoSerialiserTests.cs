using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Reflection;
using System.Xml.Serialization;
using ApprovalTests;
using CommonTestUtils;
using FluentAssertions;
using GitUI.Script;
using NUnit.Framework;

namespace GitUITests.Script
{
    [TestFixture]
    public class ScriptInfoSerialiserTests
    {
        private ScriptInfoSerialiser _scriptInfoSerialiser;

        [SetUp]
        public void Setup()
        {
            _scriptInfoSerialiser = new ScriptInfoSerialiser();
        }

        [TestCase(null)]
        [TestCase("")]
        [TestCase("\t")]
        public void Deserialise_should_return_default_scripts_if_null_or_whitespace(string xml)
        {
            var scripts = _scriptInfoSerialiser.Deserialise(xml);

            scripts.Should().BeEquivalentTo(_scriptInfoSerialiser.GetTestAccessor().GetDefaultScripts());
        }

        [Test]
        public void Deserialise_should_deserialise()
        {
            var content = EmbeddedResourceLoader.Load(Assembly.GetExecutingAssembly(), $"{GetType().Namespace}.MockData.OwnScripts.xml");

            var scripts = _scriptInfoSerialiser.Deserialise(content);

            using (var sw = new StringWriter())
            {
                var serializer = new XmlSerializer(typeof(List<ScriptInfo>));
                serializer.Serialize(sw, scripts);

                Approvals.VerifyXml(sw.ToString());
            }
        }

        [Test]
        public void Serialise_should_throw_if_null()
        {
            ((Action)(() => _scriptInfoSerialiser.Serialise(null))).Should().Throw<ArgumentNullException>();
        }

        [Test]
        public void Serialise_should_return_null_if_exception()
        {
            // No idea how to test...
            Assert.Inconclusive();
        }

        [Test]
        public void Serialise_should_serialise()
        {
            Approvals.VerifyXml(_scriptInfoSerialiser.Serialise(GetDefaultScripts()));
        }

        private IList<ScriptInfo> GetDefaultScripts() => new Collection<ScriptInfo>
        {
            new ScriptInfo
            {
                HotkeyCommandIdentifier = 9000,
                Name = "Fetch changes after commit",
                Command = "git",
                Arguments = "fetch",
                RunInBackground = false,
                AskConfirmation = true,
                OnEvent = ScriptEvent.AfterCommit,
                AddToRevisionGridContextMenu = false,
                Enabled = false
            },
            new ScriptInfo
            {
                HotkeyCommandIdentifier = 9001,
                Name = "Update submodules after pull",
                Command = "git",
                Arguments = "submodule update --init --recursive",
                RunInBackground = false,
                AskConfirmation = true,
                OnEvent = ScriptEvent.AfterPull,
                AddToRevisionGridContextMenu = false,
                Enabled = false
            }
        };
    }
}
