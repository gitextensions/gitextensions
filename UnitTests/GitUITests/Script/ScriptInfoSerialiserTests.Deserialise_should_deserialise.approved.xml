<ArrayOfScriptInfo xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <ScriptInfo>
    <Enabled>false</Enabled>
    <Name>Fetch changes after commit</Name>
    <Command>git</Command>
    <Arguments>fetch</Arguments>
    <AddToRevisionGridContextMenu>false</AddToRevisionGridContextMenu>
    <OnEvent>AfterCommit</OnEvent>
    <AskConfirmation>true</AskConfirmation>
    <RunInBackground>false</RunInBackground>
    <IsPowerShell>false</IsPowerShell>
    <HotkeyCommandIdentifier>9000</HotkeyCommandIdentifier>
    <Icon>bug</Icon>
  </ScriptInfo>
  <ScriptInfo>
    <Enabled>false</Enabled>
    <Name>Update submodules after pull</Name>
    <Command>git</Command>
    <Arguments>submodule update --init --recursive</Arguments>
    <AddToRevisionGridContextMenu>false</AddToRevisionGridContextMenu>
    <OnEvent>AfterPull</OnEvent>
    <AskConfirmation>true</AskConfirmation>
    <RunInBackground>false</RunInBackground>
    <IsPowerShell>false</IsPowerShell>
    <HotkeyCommandIdentifier>9001</HotkeyCommandIdentifier>
    <Icon>bug</Icon>
  </ScriptInfo>
  <ScriptInfo>
    <Enabled>false</Enabled>
    <Name>Example</Name>
    <Command>c:\windows\system32\calc.exe</Command>
    <Arguments />
    <AddToRevisionGridContextMenu>false</AddToRevisionGridContextMenu>
    <OnEvent>ShowInUserMenuBar</OnEvent>
    <AskConfirmation>false</AskConfirmation>
    <RunInBackground>false</RunInBackground>
    <IsPowerShell>false</IsPowerShell>
    <HotkeyCommandIdentifier>9002</HotkeyCommandIdentifier>
    <Icon>bug</Icon>
  </ScriptInfo>
  <ScriptInfo>
    <Enabled>false</Enabled>
    <Name>Open on GitHub</Name>
    <Command>{openurl}</Command>
    <Arguments>https://github.com{cDefaultRemotePathFromUrl}/commit/{sHash}</Arguments>
    <AddToRevisionGridContextMenu>true</AddToRevisionGridContextMenu>
    <OnEvent>None</OnEvent>
    <AskConfirmation>false</AskConfirmation>
    <RunInBackground>false</RunInBackground>
    <IsPowerShell>false</IsPowerShell>
    <HotkeyCommandIdentifier>9003</HotkeyCommandIdentifier>
    <Icon>bug</Icon>
  </ScriptInfo>
  <ScriptInfo>
    <Enabled>false</Enabled>
    <Name>Fetch All Submodules</Name>
    <Command>git</Command>
    <Arguments>submodule foreach --recursive git fetch --all</Arguments>
    <AddToRevisionGridContextMenu>true</AddToRevisionGridContextMenu>
    <OnEvent>None</OnEvent>
    <AskConfirmation>false</AskConfirmation>
    <RunInBackground>false</RunInBackground>
    <IsPowerShell>false</IsPowerShell>
    <HotkeyCommandIdentifier>9004</HotkeyCommandIdentifier>
    <Icon>bug</Icon>
  </ScriptInfo>
  <ScriptInfo>
    <Enabled>true</Enabled>
    <Name>test</Name>
    <Command>notepad</Command>
    <Arguments />
    <AddToRevisionGridContextMenu>false</AddToRevisionGridContextMenu>
    <OnEvent>None</OnEvent>
    <AskConfirmation>false</AskConfirmation>
    <RunInBackground>false</RunInBackground>
    <IsPowerShell>false</IsPowerShell>
    <HotkeyCommandIdentifier>9005</HotkeyCommandIdentifier>
    <Icon>bug</Icon>
  </ScriptInfo>
</ArrayOfScriptInfo>